File: backend/app/schemas/superadmin.py
"""Pydantic schemas for superadmin platform management."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import RestaurantPlan

RestaurantPlanLiteral = Literal["basic", "starter", "custom"]


class RestaurantCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Annotated[str, Field(min_length=1, max_length=255)]
    slug: Annotated[str, Field(min_length=1, max_length=100, pattern=r'^[a-z0-9-]+$')]
    admin_email: Annotated[str, Field(min_length=1, max_length=255)]
    admin_password: Annotated[str, Field(min_length=8, max_length=100)]


class RestaurantUpdate(BaseModel):
    """Partial update. Precedence when both plan and bools are sent:
    1) apply plan + its four preset bools
    2) apply any explicit bool overrides on top (explicit wins).
    """

    model_config = ConfigDict(extra="forbid")

    name: Annotated[str, Field(min_length=1, max_length=255)] | None = None
    is_active: bool | None = None
    plan: RestaurantPlanLiteral | None = None
    order_enabled: bool | None = None
    call_waiter_enabled: bool | None = None
    ar_enabled: bool | None = None
    qr_pay_enabled: bool | None = None


class AdminEmailUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    email: Annotated[str, Field(min_length=1, max_length=255)]


class AdminInfo(BaseModel):
    id: uuid.UUID
    email: str


class RestaurantResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, use_enum_values=True)

    id: uuid.UUID
    name: str
    slug: str
    is_active: bool
    plan: RestaurantPlan
    order_enabled: bool
    call_waiter_enabled: bool
    ar_enabled: bool
    qr_pay_enabled: bool
    created_at: datetime
    updated_at: datetime
    admins: list[AdminInfo] = []


class RestaurantCreateResponse(BaseModel):
    restaurant: RestaurantResponse
    admin_email: str

File: frontend/src/lib/schemas/superadmin.ts
import { z } from 'zod'

const adminInfoSchema = z.object({
  id: z.string().uuid(),
  email: z.string(),
})

export const restaurantPlanSchema = z.enum(['basic', 'starter', 'custom'])
export type RestaurantPlan = z.infer<typeof restaurantPlanSchema>

/** Defaults written when a plan is assigned (mirrors backend apply_plan_preset). */
export function planPreset(plan: RestaurantPlan): {
  order_enabled: boolean
  call_waiter_enabled: boolean
  ar_enabled: boolean
  qr_pay_enabled: boolean
} {
  if (plan === 'basic') {
    return {
      order_enabled: false,
      call_waiter_enabled: false,
      ar_enabled: false,
      qr_pay_enabled: false,
    }
  }
  if (plan === 'starter') {
    return {
      order_enabled: true,
      call_waiter_enabled: true,
      ar_enabled: false,
      qr_pay_enabled: false,
    }
  }
  return {
    order_enabled: true,
    call_waiter_enabled: true,
    ar_enabled: true,
    qr_pay_enabled: true,
  }
}

export const restaurantResponseSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  slug: z.string(),
  is_active: z.boolean(),
  plan: restaurantPlanSchema,
  order_enabled: z.boolean(),
  call_waiter_enabled: z.boolean(),
  ar_enabled: z.boolean(),
  qr_pay_enabled: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
  admins: z.array(adminInfoSchema).default([]),
})

export const restaurantCreateResponseSchema = z.object({
  restaurant: restaurantResponseSchema,
  admin_email: z.string(),
})

export type RestaurantResponse = z.infer<typeof restaurantResponseSchema>
export type RestaurantCreateResponse = z.infer<typeof restaurantCreateResponseSchema>
